Feature monte carlo collisions

roelof-groenewald opened the pull request:
Implementation of particle-neutral collisions for elastic scattering, ionization and excitation events.

peterscherpelz committed:
Update copyright notices

roelof-groenewald committed:
allow specification of boundary potentials at runtime when using Dirichlet boundary conditions in the electrostatic solver (labframe)

roelof-groenewald committed:
added parsing to boundary potentials specified at runtime to allow time dependence through a mathematical expression with t (time)

roelof-groenewald committed:
updated to picmistandard 0.0.14 in order to set the electrostatic solver convergence threshold

roelof-groenewald committed:
update docs

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_dirichletbcs

roelof-groenewald committed:
Merge branch 'development' into feature_dirichletbcs

roelof-groenewald committed:
Initial commit of MCC development. Collision type background_mcc handles collisions with a neutral background species

roelof-groenewald committed:
added back scattering and started expanding the multiple scattering processes functionality some

roelof-groenewald committed:
various changes requested during PR review

PhilMiller commented on Python/pywarpx/picmi.py:
```diff
@@ -737,10 +743,12 @@ def initialize_inputs(self):
-                                              self.injection_plane_normal_vectors[i])
+            self.species[i].initialize_inputs(
+                self.layouts[i],
+                False if self.initialize_self_fields[i] is None else True,
```
Without digging too deeply into the code, I'm slightly worried on reading this that it messes with the potential case of `self.initialize_self_fields[i] == False`

Should the whole argument expression instead read

`self.initialize_self_fields[i] is not None and self.initialize_self_fields[i]`

PhilMiller commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -346,8 +350,14 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
+    // This is apparently necessary for how AMReX expects phi
+    for (int lev=0; lev < rho.size(); lev++){
+        phi[lev]->mult(-1.*PhysConst::ep0);
+    }
```
It would be best to fold this into the assignment expression, so that the data don't have to be streamed over twice

PhilMiller commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -358,6 +368,122 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
+    }} // lev & MFIter
+
+    /*
+    Wei-Ting's implementation with GPU acceleration:
```
Not that this is live code, but it looks like it assigns values for the *whole* domain, and not just the boundary cells. Besides being locally inefficient, it could hypothetically mess with the iterative solver, since it potentially replaces the most recent solution with the straight linear interpolation, making it start to converge from scratch.

PhilMiller commented on Source/Particles/Collision/MCCScattering.H:
```diff
@@ -0,0 +1,73 @@
+{
+    // generate random unit vector for the new velocity direction
+    amrex::Real theta = amrex::Random() * 2.0 * MathConst::pi;
+    amrex::Real z = 2.0 * amrex::Random() - 1.0;
```
These calls need to take `engine` as an argument

PhilMiller commented on Source/Particles/Collision/MCCScattering.H:
```diff
@@ -0,0 +1,73 @@
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void ElasticScattering ( amrex::ParticleReal& ux, amrex::ParticleReal& uy,
+                         amrex::ParticleReal& uz, amrex::ParticleReal& uCOM_x,
+                         amrex::ParticleReal& uCOM_y, amrex::ParticleReal& uCOM_z,
```
The `uCOM` variables could be passed more cheaply as `const amrex::ParticleReal` instead of by reference.

PhilMiller commented on Source/Particles/Collision/BackgroundMCCCollision.cpp:
```diff
@@ -0,0 +1,235 @@
+
+                    // check if this collision should be performed and call
+                    // the appropriate scattering function
+                    if (col_select < nu_i){
```
As noted, this needs a `break` - it's tired out! ;-)

PhilMiller commented on Source/Particles/Collision/BackgroundMCCCollision.H:
```diff
@@ -0,0 +1,53 @@
+#include "CollisionBase.H"
+#include <AMReX_ParmParse.H>
+
+class BackgroundMCCCollision
```
Does the existing code put functionality like this in some sort of `namespace`?

roelof-groenewald replied:
The most similar existing code is in `Source/Particles/Collision/PairWiseCoulombCollision.H` and that doesn't use a specific `namespace`.

PhilMiller commented on Source/Particles/Collision/MCCScattering.H:
```diff
@@ -0,0 +1,73 @@
+ */
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void ElasticScattering ( amrex::ParticleReal& ux, amrex::ParticleReal& uy,
```
Does the WarpX coding style prefer a file per class, such that there would be separate `ElasticScattering.H` and `BackScattering.H`?

roelof-groenewald replied:
Possibly, but it seems to be that the functionality in individual scattering classes is too simple to justify breaking into a new file, for example the whole backscattering class only contains 3 lines. I'd say let's keep it as is until someone complains when we try to push it upstream.

PhilMiller commented on Source/Particles/Collision/MCCScattering.H:
```diff
@@ -0,0 +1,73 @@
+    amrex::Real z = 2.0 * amrex::Random() - 1.0;
+    amrex::Real xy = std::sqrt(1 - z*z);
+    amrex::Real x = xy * std::cos(theta);
+    amrex::Real y = xy * std::sin(theta);
```
Maybe it's common knowledge, but a comment giving a reference for this randomized velocity direction formula would be helpful for non-physicists like me.

PhilMiller commented on Source/Particles/Collision/MCCScattering.H:
```diff
@@ -0,0 +1,73 @@
+/** \brief Function to perform elastic scattering of a particle in the lab
+ * frame. The particle velocities transformed to the COM frame where a hard
+ * sphere collision occurs. The resulting particle velocities are transformed
+ * back to the lab frame and the input particle's velocity is updated.
```
If there's any sort of reference for the overall algorithm, this would be a good place to cite it.

roelof-groenewald replied:
Thanks for this comment. I'm planning to add a reference to the documentation explaining the MCC algorithm along with all the parameters involved in it.

roelof-groenewald committed:
added charge exchange collision handling

roelof-groenewald committed:
added CrossSectionHandler class to install collision process cross-section calculators

roelof-groenewald committed:
fixed issue causing old tests to break and added a new test for time varying boundary potentials

roelof-groenewald committed:
added file reading for cross-section data

roelof-groenewald committed:
added input parameter for energy lost during inelastic collisions and changed how secondary species are passed for ionization events

roelof-groenewald committed:
possibly a fix for the failed time varying boundary condition test

roelof-groenewald committed:
changed permission on the analysis file for the time varying BCs test

roelof-groenewald committed:
switched to using yt for data analysis since h5py is not available

PhilMiller commented on Source/Utils/WarpXConst.H:
```diff
@@ -33,6 +33,8 @@ namespace PhysConst
     static constexpr auto xi_c2 = static_cast<amrex::Real>( 1.1728865132395492e-35 ); // This should be usable for single precision, though
     // very close to smallest number possible (1.2e-38)
+
+    static constexpr auto kb   = static_cast<amrex::Real>( 1.38064852e-23 );
```
Not that it'll be material, but the site linked in the comment above has a slightly different value, rounded by a few digits from this one.

PhilMiller replied:
https://physics.nist.gov/cgi-bin/cuu/Value?k

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_dirichletbcs

roelof-groenewald committed:
made changes compatible with PR#1730; changed potential boundary setting routine to use the ParallelFor construct and set all boundaries in a single call

roelof-groenewald committed:
fixed typo in computePhiRZ

roelof-groenewald committed:
updated docs and fixed other minor typos

roelof-groenewald committed:
added ionization - requires work to add to the amrex::ParallelForRNG loop

roelof-groenewald committed:
switched the MCC ionization handling to use the same workflow as other particle creation processes i.e. using the FilterCopyTransform functionality

roelof-groenewald committed:
updated the docs with the input parameters needed to include MCC in a run

roelof-groenewald committed:
added test for MCC and a function to ensure that cross-section data is provided with equal energy steps

roelof-groenewald committed:
fixed issue with build failing when USE_OMP=TRUE and some of the naming issues in Examples/Physics_applications/capacitive_discharge but I am not sure what to do about the other files in that directory

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_dirichletbcs

PhilMiller committed:
Improve file name construction to be strictly C++ compliant

roelof-groenewald committed:
fixed bug in returning from loop over dimensions when setting boundary potentials rather than continuing

roelof-groenewald committed:
changed to setting potentials on domain boundaries rather than tilebox boundaries and changed picmi.py to accept boundary potentials

roelof-groenewald committed:
now using domain.surroundingNodes() to get the proper boundary cells for the physical domain

roelof-groenewald committed:
fixed typo in variable name specifying z-boundary potential

PhilMiller committed:
WIP GPU Support

ax3l committed:
Fix QED Build (CUDA 10.0)

Replace capture of a host-side array with unnamed members for E & B
field transport with a nicely named struct that transports the
Array4's as members.

This is harder to mix up and thus more self-documenting and solves an
issue with NVCC 10.0 of the form:
```
nvcc_internal_extended_lambda_implementation: In instantiation of '__nv_dl_wrapper_t<Tag, F1, F2, F3, F4, F5>::__nv_dl_wrapper_t(F1, F2, F3, F4, F5) [with Tag = __nv_dl_tag<int (*)(amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>&, amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>&, amrex::Box, const amrex::Array4<const double> (&)[6], int, int, const SchwingerFilterFunc&, const SmartCreate&, const SmartCreate&, const SchwingerTransformFunc&), filterCreateTransformFromFAB<1, amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>, amrex::Array4<const double> [6], int, const SchwingerFilterFunc&, const SmartCreate&, const SmartCreate&, const SchwingerTransformFunc&>, 1>; F1 = amrex::Array4<double>; F2 = const SchwingerFilterFunc; F3 = const amrex::Array4<const double> [6]; F4 = const amrex::Box; F5 = int*]':
/home/ubuntu/repos/WarpX/Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H:174:28:   required from 'Index filterCreateTransformFromFAB(DstTile&, DstTile&, amrex::Box, const FABs&, Index, Index, FilterFunc&&, CreateFunc1&&, CreateFunc2&&, TransFunc&&) [with int N = 1; DstTile = amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>; FABs = amrex::Array4<const double> [6]; Index = int; FilterFunc = const SchwingerFilterFunc&; CreateFunc1 = const SmartCreate&; CreateFunc2 = const SmartCreate&; TransFunc = const SchwingerTransformFunc&]'
/home/ubuntu/repos/WarpX/Source/Particles/MultiParticleContainer.cpp:1169:167:   required from here
nvcc_internal_extended_lambda_implementation:70:103: error: invalid initializer for array member 'const amrex::Array4<const double> __nv_dl_wrapper_t<__nv_dl_tag<int (*)(amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>&, amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>&, amrex::Box, const amrex::Array4<const double> (&)[6], int, int, const SchwingerFilterFunc&, const SmartCreate&, const SmartCreate&, const SchwingerTransformFunc&), filterCreateTransformFromFAB<1, amrex::ParticleTile<0, 0, 4, 0, amrex::ArenaAllocator>, amrex::Array4<const double> [6], int, const SchwingerFilterFunc&, const SmartCreate&, const SmartCreate&, const SchwingerTransformFunc&>, 1>, amrex::Array4<double>, const SchwingerFilterFunc, const amrex::Array4<const double> [6], const amrex::Box, int*>::f3 [6]'
```

PhilMiller committed:
CUDA: Quiet numerous warnings about unused-variable-warning suppressions being unreachable statements

PhilMiller committed:
Compiles on GPU; may even run as intended

PhilMiller committed:
Delete overwrought attempt at polymorphic implementation

PhilMiller committed:
Fix compilation error from nvcc being stupid

roelof-groenewald committed:
fixed improper input file for MCC test and updated reference results - a statistical test of the MCC routine would be better so that reference results should not change with changes in the RNG

PhilMiller committed:
Runs on CPU and GPU now

PhilMiller committed:
Clean up GPU-related memory/allocation management and function usage

PhilMiller committed:
Try inlining MCCProcess::getCrossSection to appease HIP and SYCL compilers

PhilMiller committed:
Fix up style/formatting issues

PhilMiller committed:
Typedef to make stuff cleaner and simpler

PhilMiller committed:
MCC: Make helper functions static

PhilMiller committed:
MCC: Pull parsing out to a helper

PhilMiller committed:
MCC: Name member variables according to convention

PhilMiller committed:
MCC: Pull out part of constructor

PhilMiller committed:
MCC: Add constructor that will take any iterable source for energies/cross sections

PhilMiller committed:
MCC: Overload operator new/delete to allocate in managed memory, to make later use more straightforward

PhilMiller committed:
MCC: Add process type for ionization

PhilMiller committed:
MCC: Expose a method for adding processes programmatically

PhilMiller committed:
MCC: Follow convention of all types being 'class', which keeps grep easy

PhilMiller committed:
MCC: Fix a formatting silliness

roelof-groenewald committed:
added a check that the collision cross-section is zero at the energy penalty for the collision to ensure that no collision will happen with a particle with insufficient energy to pay the energy cost

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_monte_carlo_collisions

roelof-groenewald committed:
updated MCC input files to new standard inputs

roelof-groenewald committed:
reverted incorrect changes that was messed up during various upstream and branch merges

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_monte_carlo_collisions

roelof-groenewald committed:
moved the MCC benchmark results to the Examples section in the documentation, which allows us to meet the style requirements - the tests are ongoing and the results will be provided in a following commit

PhilMiller committed:
Add GPU synchronization after collisions

roelof-groenewald committed:
added benchmark results and updated test results with the refined cross-sections needed to accurately calculate the benchmark cases

roelof-groenewald commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Monte Carlo Collisions (C++)](https://trello.com/c/olZQGG3O/16-monte-carlo-collisions-c)

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_monte_carlo_collisions

roelof-groenewald committed:
removed example input files for benchmarks since the style requirement prohibits input files not included in an automated test; also updated the reference results for the MCC test which changed slightly after merging upstream development and updated amrex

PhilMiller committed:
CLean up indentation and bit of commented out code

PhilMiller committed:
Inline addProcess method and refactor

roelof-groenewald committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

roelof-groenewald committed:
switched MCC files copyright to Modern Electron

Commit:
Remove sync calls, which are unnecessary on support modern hardware

roelof-groenewald committed:
removed He collision cross-sections and instead the new warpx-data repository to access those files; also added a call in run_test.sh to clone the new repository during tests

roelof-groenewald committed:
Apply suggestions from code review

Co-authored-by: David Grote <dpgrote@lbl.gov>
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

roelof-groenewald committed:
cleaned up the MCC documentation a bit

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into feature_monte_carlo_collisions

roelof-groenewald committed:
added include statements now needed by the MCC (after recent PR merges) and updated the MCC test reference values which changed slightly due to changing the value of Boltzmann's constant
